python/lsst/pipe/tasks/characterizeImage.py

```diff
@@ -0,0 +1,493 @@
+                dataRef.put(charRes.background, "icExpBackground")
+
+        return charRes
+
+    @pipeBase.timeMethod
+    def characterize(self, exposure, exposureIdInfo, background=None):
+        """!Characterize a science image
+
+        Peforms the following operations:
+        - Iterate the following config.psfIterations times, or once if config.doMeasurePsf false:
+            - detect and measure sources and estimate PSF (see detectMeasureAndEstimatePsf for details)
+        - interpolate over cosmic rays
+        - perform final measurement
+
+        @param[in,out] exposure  exposure to characterize (an lsst.afw.image.ExposureF or similar).
+            The following changes are made:
+            - update or set psf
+            - set apCorrMap
+            - update detection and cosmic ray mask planes
+            - subtract background and interpolate over cosmic rays
+        @param[in] exposureIdInfo  ID info for exposure (an lsst.daf.butlerUtils.ExposureIdInfo)
+        @param[in] background  model of background model already subtracted from exposure
+            (an lsst.afw.math.BackgroundList). May be None if no background has been subtracted,
+            which is typical for image characterization.
+
+        @return pipe_base Struct containing these fields, all from the final iteration
+        of detectMeasureAndEstimatePsf:
+        - exposure: characterized exposure; image is repaired by interpolating over cosmic rays,
+            mask is updated accordingly, and the PSF model is set
+        - sourceCat: detected sources (an lsst.afw.table.SourceCatalog)
+        - background: model of background subtracted from exposure (an lsst.afw.math.BackgroundList)
+        - psfCellSet: spatial cells of PSF candidates (an lsst.afw.math.SpatialCellSet)
+        """
+        self._frame = self._initialFrame # reset debug display frame
+
+        if not self.config.doMeasurePsf and not exposure.hasPsf():
+            raise pipeBase.TaskError("exposure has no PSF model and config.doMeasurePsf false")
+
+        if background is None:
+            background = BackgroundList()
+
+        # make a deep copy of the mask
+        originalMask = exposure.getMaskedImage().getMask().clone()
+
+        # subtract an initial estimate of background level
+        estBg = estimateBackground(
+            exposure = exposure,
+            backgroundConfig = self.config.background,
+            subtract = False, # this makes a deep copy, which we don't want
+        )[0]
+        image = exposure.getMaskedImage().getImage()
+        image -= estBg.getImageF()
+        background.append(estBg)
+
+        psfIterations = self.config.psfIterations if self.config.doMeasurePsf else 1
+        for i in range(psfIterations):
+            if i > 1:
+                # restore original mask so that detections and cosmic rays
+                # are only marked by the final iteration
+                exposure.getMaskedImage().getMask()[:] = originalMask
```

Comment:
If we know which bits are set (and we should) we can clear them instead of saving a copy of the mask.  More memory and cache efficient; I think you can say `exposure.getMaskedImage().getMask()[:] &= ~bitsToClear`